src/transform/src/analysis.rs

```diff
@@ -568,12 +577,81 @@ mod types {
                 .map(|child| &results[child])
                 .collect::<Vec<_>>();
 
-            if offsets.iter().all(|o| o.is_some()) {
-                let input_cols = offsets.into_iter().rev().map(|o| o.as_ref().unwrap());
-                let subtree_column_types = expr.try_col_with_input_cols(input_cols);
-                subtree_column_types.ok()
-            } else {
-                None
+            // For most expressions we'll apply `try_col_with_input_cols`, but for `Get` expressions
+            // we'll want to combine what we know (iteratively) with the stated `Get::typ`.
+            match expr {
+                MirRelationExpr::Get {
+                    id: mz_expr::Id::Local(i),
+                    typ,
+                    ..
+                } => {
+                    let mut result = typ.column_types.clone();
+                    if let Some(o) = depends.bindings().get(i) {
+                        if let Some(t) = results.get(*o) {
```

Comment:
I'll need to check. One of these tests is "are you running in pessimistic mode" which does need to work and not panic. And .. I'm not certain how careful pessimistic mode is in setting things up, because it does not expect folks to look around too much.